File: Coding_1124 Backtracking/이강훈/최원준/3번.py
# https://school.programmers.co.kr/learn/courses/30/lessons/43164

'''
1. 아이디어 :
    중복되는 티켓이 있다는 얘기가 없어서 해시맵으로 했다가, 중복되는 티켓이 존재해서 갈아 엎었다.....
    used 배열을 만들어서 사용한 티켓을 추적한다.
    path의 길이가 tickets의 길이+1 과 같으면 모든 도착지를 방문했으므로, path를 저장한다.
    티켓을 돌면서, 사용 안한 티켓중, 티켓의 시작점이 마지막 방문지와 같으면 재귀를 돌린다.
    마지막으로 ans를 정렬하여 제일 앞의 값을 리턴
2. 시간복잡도 :
    O(n!)
3. 자료구조 :
    배열
'''


def solution(tickets):
    def backtrack(start, path):
        if len(path) == len(tickets) + 1:
            ans.append(path.copy())

        for i in range(len(tickets)):
            if not used[i] and tickets[i][0] == start:
                used[i] = True
                path.append(tickets[i][1])

                backtrack(tickets[i][1], path)

                path.pop()
                used[i] = False

    used = [False] * len(tickets)
    ans = []

    backtrack("ICN", ["ICN"])
    ans.sort()

    return ans[0]

File: Coding_231222 2D_DP/이강훈/이강훈/3번.java
/*
https://school.programmers.co.kr/learn/courses/30/lessons/42898#

아이디어
우측, 하단 합

자료구조
배열

시간복잡도
O(NM)
*/

public class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int[][] map = new int[n][m];

        for (int[] puddle : puddles) {
            map[puddle[1] - 1][puddle[0] - 1] = -1;
        }

        map[0][0] = 1;

        for (int i = 0; i < map.length; i++) {
            for (int j = 0; j < map[0].length; j++) {
                if (map[i][j] == -1) continue;

                if (j + 1 < m && map[i][j + 1] != -1) {
                    map[i][j + 1] += map[i][j];
                    map[i][j + 1] %= 1000000007;
                }

                if (i + 1 < n && map[i + 1][j] != -1) {
                    map[i + 1][j] += map[i][j];
                    map[i + 1][j] %= 1000000007;
                }
            }
        }

        return map[n - 1][m - 1];
    }
}
